diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/kdoc/CommentsFormatting.kt

```diff
@@ -0,0 +1,255 @@
+                COMMENT_WHITE_SPACE.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+                    node.treeParent.addChild(PsiWhiteSpaceImpl(" ".repeat(configuration.maxSpaces)), node)
+                }
+            } else if (node.treePrev.text.length != configuration.maxSpaces) {
+                COMMENT_WHITE_SPACE.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+                    (node.treePrev as LeafPsiElement).replaceWithText(" ".repeat(configuration.maxSpaces))
+                }
+            }
+        }
+    }
+
+    private fun checkWhiteSpaceBeforeComment(node: ASTNode) {
+        if (node.text.takeWhile { it == '/' || it == ' ' || it == '*' }.length == APPROPRIATE_COMMENT_SPACES)
+            return
+
+        COMMENT_WHITE_SPACE.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+            val commentText = node.text.drop(2).trim()
+
+            (node as LeafPsiElement).replaceWithText("// $commentText")
+        }
+    }
+
+    private fun checkClassComment(node: ASTNode) {
+        if (isFirstComment(node)) {
+            if (node.treeParent.elementType == BLOCK || node.treeParent.elementType == CLASS_BODY)
+                checkFirstCommentSpaces(node)
+            else
+                checkFirstCommentSpaces(node.treeParent)
+
+            return
+        }
+
+        if (!node.treeParent.treePrev.isWhiteSpace()) {
+            COMMENT_NEW_LINES.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+                node.treeParent.treeParent.addChild(PsiWhiteSpaceImpl("\n"), node.treeParent)
+            }
+        } else {
+            if (node.treeParent.treePrev.text.countSubStringOccurrences("\n") == 1 || node.treeParent.treePrev.text.countSubStringOccurrences("\n") > 2) {
+                COMMENT_NEW_LINES.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+                    (node.treeParent.treePrev as LeafPsiElement).replaceWithText("\n\n")
+                }
+            }
+        }
+    }
+
+    private fun checkFirstCommentSpaces(node: ASTNode) {
+        if (node.treePrev.isWhiteSpace()) {
+            if (node.treePrev.text.countSubStringOccurrences("\n") > 1) {
+                FIRST_COMMENT_NO_SPACES.warnAndFix(configRules, emitWarn, isFixMode, node.text, node.startOffset) {
+                    (node.treePrev as LeafPsiElement).replaceWithText("\n")
+                }
+            }
+        }
+    }
+
+
+    private fun isFirstComment(node: ASTNode): Boolean {
+        if (node.treeParent.elementType == BLOCK || node.treeParent.elementType == CLASS_BODY) {
+            return if (node.treePrev.isWhiteSpace())
+                node.treePrev.treePrev.elementType == LBRACE
+            else
+                node.treePrev.elementType == LBRACE
+        }
+
+        return node.treeParent.treePrev.treePrev.elementType == LBRACE
```

Comment:
Once again, how safe and reliable is this triple `treePrev` call? Also, please add a comment explaining what does this node correspond to in this case.